pkg/controller/framework/shared_informer.go

```diff
@@ -0,0 +1,243 @@
+}
+
+func (v *votingController) Run(stopCh <-chan struct{}) {
+	v.informer.runVotes.Done()
+}
+
+func (v *votingController) HasSynced() bool {
+	return v.informer.HasSynced()
+}
+
+type shareableProcessor struct {
+	listeners []*processorListener
+}
+
+const (
+	channelDepth = 1000
+)
+
+type processorListener struct {
+	notifications chan interface{}
+
+	handler ResourceEventHandler
+}
+
+type updateNotification struct {
+	oldObj interface{}
+	newObj interface{}
+}
+
+type addNotification struct {
+	newObj interface{}
+}
+
+type deleteNotification struct {
+	oldObj interface{}
+}
+
+func NewProcessListener(handler ResourceEventHandler) *processorListener {
+	return &processorListener{
+		notifications: make(chan interface{}, channelDepth),
+		handler:       handler,
+	}
+}
+
+func (s *ShareableInformer) Run(stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+
+	func() {
+		s.startedLock.Lock()
+		defer s.startedLock.Unlock()
+		s.started = true
+	}()
+
+	s.runVotes.Wait()
+
+	s.processor.Run(stopCh)
+	s.controller.Run(stopCh)
+}
+
+func (s *ShareableInformer) isStarted() bool {
+	s.startedLock.Lock()
+	defer s.startedLock.Unlock()
+	return s.started
+}
+
+func (s *ShareableInformer) HasSynced() bool {
+	return s.controller.HasSynced()
+}
+
+func (s *ShareableInformer) GetStore() cache.Store {
+	return s.store
+}
+
+func (s *ShareableInformer) GetController() ControllerInterface {
+	return &votingController{informer: s}
+}
+
+func (s *ShareableInformer) AddEventHandler(handler ResourceEventHandler) error {
+	s.startedLock.Lock()
+	defer s.startedLock.Unlock()
+
+	if s.started {
+		return fmt.Errorf("informer has already started")
+	}
+
+	// after adding your event handler, you'll need to call `Run` on your ControllerInterface to count your vote
+	s.runVotes.Add(1)
+
+	listener := NewProcessListener(handler)
+	s.processor.listeners = append(s.processor.listeners, listener)
+	return nil
+}
+
+func (s *ShareableInformer) HandleDeltas(obj interface{}) error {
+	// from oldest to newest
+	for _, d := range obj.(cache.Deltas) {
+		switch d.Type {
+		case cache.Sync, cache.Added, cache.Updated:
+			if old, exists, err := s.store.Get(d.Object); err == nil && exists {
+				if err := s.store.Update(d.Object); err != nil {
+					return err
+				}
+				s.processor.distribute(updateNotification{oldObj: old, newObj: d.Object})
+			} else {
+				if err := s.store.Add(d.Object); err != nil {
+					return err
+				}
+				s.processor.distribute(addNotification{newObj: d.Object})
+			}
+		case cache.Deleted:
+			if err := s.store.Delete(d.Object); err != nil {
+				return err
+			}
+			s.processor.distribute(deleteNotification{oldObj: d.Object})
+		}
+	}
+	return nil
+}
+
+// TODO get more sophisticated about handling stuck controllers, but prove that this works for now.
+// Probably handle it with an infinite slice, an unbuffered channel, and a go routine
+func (p *shareableProcessor) distribute(obj interface{}) {
+	for _, listener := range p.listeners {
```

Comment:
> yeah.. the problem is that a single stuck controller may block all other controllers, which is bad.
> I don't think infinite slice is a good idea, as this may lead to OOMs.

We use infinite slices today, so this wouldn't be a new problem.  A stuck controller will block its incoming queue, which is an unbounded slice.  This just made the problem a little more obvious.

Would you be ok with an unbounded slice implementation to start and we can later think through a re-join scenario?  The first implementation that occurs to me requires a slice as large as the existing cache and a full lock while while we prime the notification list.